### PR TITLE
Implement `forc doc`

### LINKED ISSUE #161: implement `forc doc`
it should generate docs for Sway constructs: Structs, Enums, Traits, Functions etc
it should work similar as `cargo doc`

### PR DESCRIPTION
**Pull request by eureka-cpu**
~~Opening up for visibility, still reading through `cargo doc` and previous comments related to this~~

Closes #161 

- [x] #3027 
- [x] #3033 
- [x] #3036 
- [x] #3038 
- [x] #3088 
- [x] #3090

### CODE REVIEW

**Review by mitchmindtree — COMMENTED** [review]
Wicked stuff @eureka-cpu, the code is looking great! Super easy to read :ok_hand: 

### `--open`

I tried running this locally on the `sway-lib-std` with `--open` and neither my browser (firefox) or file manager (GNOME Files) opened. Not a big issue, I'd even be happy to leave the open functionality for a future PR, but just a heads up!

### Module level docs / navigation

Also, I couldn't see an `index.html` or any module HTML files in `out/doc` that would allow me to navigate between the different modules and the items they contain. Are we aiming to include this in this PR or are we leaving this for follow-up?

### ci check

It would be great to have a CI pass that runs `forc doc` on at `sway-lib-core`, `sway-lib-std` and the `examples`.

**Comment by eureka-cpu** [comment]
> Wicked stuff @eureka-cpu, the code is looking great! Super easy to read 👌
> ### `--open`
> 
> I tried running this locally on the `sway-lib-std` with `--open` and neither my browser (firefox) or file manager (GNOME Files) opened. Not a big issue, I'd even be happy to leave the open functionality for a future PR, but just a heads up!
> ### Module level docs / navigation
> 
> Also, I couldn't see an `index.html` or any module HTML files in `out/doc` that would allow me to navigate between the different modules and the items they contain. Are we aiming to include this in this PR or are we leaving this for follow-up?
> ### ci check
> 
> It would be great to have a CI pass that runs `forc doc` on at `sway-lib-core`, `sway-lib-std` and the `examples`.

Currently working on getting all of the index stuff finished so that it will open (:

File: forc-plugins/forc-doc/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    /// file in the current directory or any parent directory.
+    #[clap(long)]
+    pub manifest_path: Option<String>,
+    /// Open the docs in a browser after building them.
+    #[clap(long)]
+    pub open: bool,
+    /// Offline mode, prevents Forc from using the network when managing dependencies.
+    /// Meaning it will only try to use previously downloaded dependencies.
+    #[clap(long = "offline")]
+    pub offline_mode: bool,
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/cli.rs:15]: These are copied from the other plugins and `forc` commands for consistency, should I change them anyway? the command itself will still be `--offline` and `--silent` or `-s`. I think it was just for code clarity that it was done this way.

File: forc-plugins/forc-doc/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    // get manifest directory
+    let dir = if let Some(ref path) = manifest_path {
+        PathBuf::from(path)
+    } else {
+        std::env::current_dir()?
+    };
+    let manifest = PackageManifestFile::from_dir(&dir)?;
+
+    // check if the out path exists
+    let out_path = PathBuf::from(&manifest.dir()).join("out");
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/main.rs:42]: I'll check, again this is just copy pasta from one of the other plugins!

File: forc-plugins/forc-doc/src/doc.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                    desc_ty,
+                } = desc
+                {
+                    docs.push(Document {
+                        module_prefix,
+                        desc_ty: *desc_ty,
+                    })
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/doc.rs:102]: They are pretty similar, I'll try to find a way to abstract it further

File: forc-plugins/forc-doc/src/cli.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mitchmindtree) [reply forc-plugins/forc-doc/src/cli.rs:15]: Yeah lets change these anyway. I think the `_mode` suffix is unnecessary on the other plugins too, but we can address those separately.

File: forc-plugins/forc-doc/src/doc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (eureka-cpu) [reply forc-plugins/forc-doc/src/doc.rs:102]: If you don't mind, I'll definitely pick up where I left off and address this in another PR. Everything is working as intended now and ready to go I believe!

**Review by JoshuaBatty — APPROVED** [review]
Great work @eureka-cpu ! The code looks super clean and readable. Happy for this to land as is and follow-up PR's be made iteratively to finish off the missing functionality.

**Review by JoshuaBatty — APPROVED** [review]

File: forc-plugins/forc-doc/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "forc-doc"
+version = "0.1.0"
```
Comment (mohammadfawaz) [inline forc-plugins/forc-doc/Cargo.toml:3]: Should this be `0.28.1`?

Comment (JoshuaBatty) [reply forc-plugins/forc-doc/Cargo.toml:3]: Yes, yes it should :) Good catch

**Review by mohammadfawaz — APPROVED** [review]
Amazing work @eureka-cpu. Super excited for this.

**Review by JoshuaBatty — APPROVED** [review]

**Comment by eureka-cpu** [comment]
Thanks for handling that for me @JoshuaBatty !

**Pull request merged**